miniserver/blockchain.js

```diff
@@ -0,0 +1,124 @@
+const walletPath = 'm/44\'/60\'/0\'/0/' // https://docs.ethers.io/v5/api/signer/#Wallet.fromMnemonic'
+
+const init = async () => {
+  Logger.log('Initializing blockchain for server')
+  try {
+    Logger.log(`config.defaultNetwork: ${config.defaultNetwork}`)
+    networkConfig = config.networks[config.defaultNetwork]
+    Logger.log(`network: ${JSON.stringify(networkConfig)}`)
+    provider = ethers.getDefaultProvider(networkConfig.url)
+    provider.pollingInterval = config.pollingInterval
+    if (networkConfig.mnemonic) {
+      for (let i = 0; i < networkConfig.numAccounts; i += 1) {
+        const path = walletPath + i.toString()
+        Logger.log(`path: ${path}`)
+        const signer = new ethers.Wallet.fromMnemonic(networkConfig.mnemonic, path)
+        signers[i] = signer.connect(provider)
+      }
+    } else {
+      signers[0] = new ethers.Wallet(networkConfig.key, networkConfig.provider)
+    }
+    for (let i = 0; i < signers.length; i += 1) {
+      Logger.log(`signers[${i}].address; ${JSON.stringify(signers[i].address)}`)
+      miniWallets[i] = new ethers.Contract(networkConfig.miniWalletAddress, MiniWallet.abi, signers[i])
+    }
+  } catch (ex) {
+    console.error(ex)
+    console.trace(ex)
+  }
+  for (const signer of signers) {
+    pendingNonces[signer.address] = 0
+    Logger.log(`[${config.defaultNetwork}][${signer.address}] Set pending nonce = 0`)
+  }
+}
+
+const sampleExecutionAddress = () => {
+  const nonces = cloneDeep(pendingNonces)
+  const probs = []
+  let sum = 0
+  for (const signer of signers) {
+    const p = 1.0 / Math.exp(nonces[signer.address])
+    probs.push(p)
+    sum += p
+  }
+  const r = Math.random() * sum
+  let s = 0
+  for (let i = 0; i < probs.length; i++) {
+    s += probs[i]
+    if (s >= r) {
+      return [i, signers[i].address, miniWallets[i]]
+    }
+  }
+  return [signers.length - 1, signers[signers.length - 1].address, miniWallets[miniWallets.length - 1]]
+}
+
+// basic executor used to send funds
+const prepareExecute = (logger = Logger.log, abortUnlessRPCError = true) => async (method, params) => {
+  const [fromIndex, from, miniWallet] = sampleExecutionAddress()
+  logger(`Sampled [${fromIndex}] ${from}`)
+  const latestNonce = await rpc.getNonce({ address: from, network: config.defaultNetwork })
+  const snapshotPendingNonces = pendingNonces[from]
+  const nonce = latestNonce + snapshotPendingNonces
+  pendingNonces[from] += 1
+  const t0 = performance.now()
+  const elapsed = () => (performance.now() - t0).toFixed(3)
+  const printNonceStats = () => `[elapsed=${elapsed()}ms][network=${config.defaultNetwork}][account=${fromIndex}][nonce=${nonce}][snapshot=${snapshotPendingNonces}][current=${pendingNonces[from]}]`
+  try {
+    logger(`[pending]${printNonceStats()}`)
+    let numAttempts = 0
+    // const f = () => {miniWallet.send(params[0], params[1], params[2])}
+    const tx = await backOff(
+      async () => miniWallet.send(params[0], params[1], params[2], {
```

Comment:
should use variable length (spread) params

Reply:
done 🙏